File: Backup Source/11-12-09/Arrows.py
'''
Created on 9 Oct 2011

@author: Elliot
'''

from params import *
import utils

class Arrows():
    def __init__(self, box, num_arrows, dt, rate_base, grad_sense, n_mem, mem_sense, onesided_flag):
        self.num_arrows = num_arrows
        self.dt = dt
        self.v = 1.0

        # Algorithm choice
        self.rates_update = self._rates_update_mem
        self.wall_handle = self._wall_aligning

        self._rs_initialise(box)
        self._vs_initialise(box)
        self._rates_initialise(rate_base, grad_sense, mem_sense, n_mem, onesided_flag)
        
        self.zeros = np.zeros_like(self.rates)
        
    def _rs_initialise(self, box):
        self.rs = np.empty([self.num_arrows, box.dim_get()], dtype=np.float)
        L_half = box.L_get() / 2.0

        for i_dim in range(box.dim_get()):
            self.rs[:, i_dim] = np.random.uniform(-L_half, +L_half,
                                                  self.num_arrows)

        i_lattice = box.i_lattice_find(self.rs)
        for i_arrow in range(self.num_arrows):
            while box.is_wall(i_lattice[i_arrow]):
                for i_dim in range(box.dim_get()):
                    self.rs[i_arrow, i_dim] = np.random.uniform(-L_half, +L_half)
                i_lattice[i_arrow] = box.i_lattice_find(np.array([self.rs[i_arrow]]))

    def _vs_initialise(self, box):
        vs_p = np.empty([self.num_arrows, box.dim_get()], dtype=np.float)
        vs_p[:, 0] = self.v
        vs_p[:, 1] = np.random.uniform(-np.pi, np.pi, self.num_arrows)
        self.vs = utils.polar_to_cart(vs_p)

    def _mem_kernel_find(self, n_mem, mem_sense):
        # Transient / Steady-state compromise parameter, 0.5 good
        A = 0.5
        N = (0.8125 * A ** 2 - 0.75 * A + 0.5) ** -0.5
        t_s = np.arange(0.0, float(n_mem) / self.rate_base, self.dt, dtype=np.float)
        g_s = self.rate_base * t_s
        self.K = (N * mem_sense * self.rate_base * np.exp(-g_s) * 
                  (1.0 - A * (g_s + (g_s ** 2.0) / 2.0)))

#        print((self.K*self.dt).sum())
#        P.plot(t_s * self.rate_base, self.K)
#        P.show()

    def _rates_initialise(self, rate_base, grad_sense, mem_sense, n_mem, onesided_flag):
        self.rates = np.empty([self.num_arrows], dtype=np.float)
        self.rate_base = rate_base        
        if self.rates_update == self._rates_update_grad:
            self.grad_sense = grad_sense
        if self.rates_update == self._rates_update_mem:
            self.heavy_flag = onesided_flag
            self._mem_kernel_find(n_mem, mem_sense)
            self.attract_mem = np.zeros([self.num_arrows, self.K.shape[0]], dtype=np.float)

    def rs_update(self, box):
        rs_test = self.rs + self.vs * self.dt
        i_obstructed, i_lattice_test = box.i_obstructed_find(rs_test)

        for i_arrow in i_obstructed:
            r_cell_test = box.r_cell_find(i_lattice_test[i_arrow])
            i_lattice_source = box.i_lattice_find(self.rs[i_arrow])

            r_source_rel = self.rs[i_arrow] - r_cell_test
            sides = np.sign(r_source_rel)
    
            delta_i_lattice = np.abs(i_lattice_source - i_lattice_test[i_arrow])
            adjacentness = delta_i_lattice.sum()

            # If test cell is same as source cell
            if adjacentness == 0:
                break
            # If test cell is directly adjacent to original
            elif adjacentness == 1:
                # Find dimension where i_lattice has changed
                dim_hit = delta_i_lattice.argmax()
            elif adjacentness == 2:
                # Dimension with largest absolute velocity component
                dim_bias = np.abs(self.vs[i_arrow]).argmax()
                i_lattice_new = i_lattice_test[i_arrow, :]
                i_lattice_new[1 - dim_bias] += sides[1 - dim_bias]
                if not box.is_wall(i_lattice_new):
                    dim_hit = 1 - dim_bias
                else:
                    dim_nonbias = 1 - dim_bias
                    i_lattice_new = i_lattice_test[i_arrow, :]
                    i_lattice_new[dim_nonbias] += sides[1 - dim_nonbias]
                    if not box.is_wall(i_lattice_new):
                        dim_hit = 1 - dim_nonbias
                    # Must be that both adjacent cells are walls
                    else:
                        dim_hit = [0, 1]
            # Change position to just off obstructing cell edge
            rs_test[i_arrow, dim_hit] = (r_cell_test + 
                                         sides * 
                                         ((box.dx_get() / 2.0) + 
                                          box.cell_buffer_get()))[dim_hit]
            self.wall_handle(i_arrow, dim_hit)

        # Note, need to extend movement to fill time-step    
        self.rs[:] = rs_test[:]

    def vs_update(self):
        self._tumble()

    def _tumble(self):
        dice_roll = np.random.uniform(0.0, 1.0, self.num_arrows)
        i_tumblers = np.where(dice_roll < (self.rates * self.dt))[0]
        vs_p = utils.cart_to_polar(self.vs[i_tumblers])
        vs_p[:, 1] = np.random.uniform(-np.pi, np.pi, i_tumblers.shape[0])
        self.vs[i_tumblers] = utils.polar_to_cart(vs_p)

    def _grad_find(self, field, dx, i_lattice):
        traj = np.asarray(np.sign(self.vs), dtype=np.int)
        for i_arrow in range(self.num_arrows):
            dim_bias = np.abs(self.vs[i_arrow]).argmax()
            traj[i_arrow, 1 - dim_bias] = 0

        grad = (field[(i_lattice + traj)[:, 0], (i_lattice + traj)[:, 1]] -
                field[(i_lattice - traj)[:, 0], (i_lattice - traj)[:, 1]]) / dx 
        return grad

    def _integral_find(self):
        return np.sum(self.attract_mem * self.K, axis=1) * self.dt

# Tumbling rate algorithms

    def _rates_update_const(self, box):
        self.rates[:] = self.rate_base

    def _rates_update_grad(self, box):
        i_lattice = box.i_lattice_find(self.rs)
        grad_attract = self._grad_find(box.attract_get(), box.dx_get(), i_lattice)
        if self.heavy_flag:
            self.rates = self.rate_base - self.grad_sense * np.maximum(grad_attract, self.zeros)
        else:
            self.rates = self.rate_base - self.grad_sense * grad_attract

    def _rates_update_mem(self, box):
        i_lattice = box.i_lattice_find(self.rs)
        self.attract_mem[:, :-1] = self.attract_mem[:, 1:]
        attract = box.attract_get()
        for i_arrow in range(self.num_arrows):
            self.attract_mem[i_arrow, 0] = attract[i_lattice[i_arrow, 0],
                                                   i_lattice[i_arrow, 1]]
        if self.heavy_flag:
            self.rates = self.rate_base * (1.0 - np.maximum(self._integral_find(), self.zeros))
        else:
            self.rates = self.rate_base * (1.0 - self._integral_find())
        print(self.rates[0])

# / Tumbling rate algorithms

# Wall handling algorithms

    def _wall_specular(self, i_arrow, dim_hit):
        self.vs[i_arrow, dim_hit] *= -1.0

    def _wall_aligning(self, i_arrow, dim_hit):
        direction = self.vs[i_arrow].copy()
        direction[dim_hit] = 0.0
        if utils.vector_mag(direction) < ZERO_THRESH:
            self.vs[i_arrow] = (random.choice([1.0, -1.0]) *
                                np.array([self.vs[i_arrow, 1],
                                          -self.vs[i_arrow, 0]]))
        else:
            self.vs[i_arrow] = utils.vector_unitise(direction) * utils.vector_mag(self.vs[i_arrow])

    def _wall_bounceback(self, i_arrow, dim_hit):
        self.vs[i_arrow] *= -1.0
        
    def _wall_stalling(self, i_arrow, dim_hit):
        self.vs[i_arrow, dim_hit] = 0.0

# / Wall handling Algorithms

File: Auxiliary Implementations/C++/src/Arrows.cpp
/*
 * Arrows.cpp
 *
 *  Created on: 3 Dec 2011
 *      Author: ejm
 */

#include "Arrows.hpp"

Arrows::Arrows(const Box& box, int num_arrows,
							 float sense_grad,
							 float t_mem, float sense_mem,
							 bool onesided_flag, char rate_alg, char bc_alg) {

	m_num_arrows = num_arrows;
	m_onesided_flag = onesided_flag;
	m_rate_alg = rate_alg;
	m_bc_alg = bc_alg;

	float run_length_base = 1.0;
	float run_time_base = 1.0;

	m_rate_base = 1.0 / run_time_base;
	m_v = run_length_base * m_rate_base;

	if (bc_alg == 'p') {
		m_pwall_handle = &Arrows::wall_specular;
	} else if (bc_alg == 'a') {
		m_pwall_handle = &Arrows::wall_aligning;
	} else if (bc_alg == 't') {
		m_pwall_handle = &Arrows::wall_stalling;
	} else if (bc_alg == 'b') {
		m_pwall_handle = &Arrows::wall_bounceback;
	}

	if (rate_alg == 'c') {
		m_prates_update = &Arrows::rates_update_const;
	} else if (rate_alg == 'g') {
		m_prates_update = &Arrows::rates_update_grad;
		m_sense_grad = sense_grad;
	} else if (rate_alg == 'm') {
		m_prates_update = &Arrows::rates_update_mem;
	} else {
		cout << "Error: Rates update mode not recognised." << endl;
	}

	rs_initialise(box);
	vs_initialise();
	rates_initialise(rate_alg, sense_mem, t_mem);
}

void Arrows::rs_update(const Box& box) {
	Array<float, 1> test_r(DIM), test_cell_r(DIM), source_r_rel(DIM), v_a(DIM), sides(DIM);
	TinyVector<int, DIM> source_i, test_i, delta_i, pos_i;
	Range i_dims_hit;

	for(int i_arrow = 0; i_arrow < m_rs.extent(firstDim); i_arrow++) {
		test_r = m_rs(i_arrow, Range::all()) + m_vs(i_arrow, Range::all()) * DELTA_t;
		test_i = box.r_to_i(test_r);
		if(box.is_wall(test_i)) {
			test_cell_r = box.i_to_r(test_i);
			source_i = box.r_to_i(m_rs(i_arrow, Range::all()));
			source_r_rel = m_rs(i_arrow, Range::all()) - test_cell_r;
			for(int i_dim = 0; i_dim < DIM; i_dim++)
				sides(i_dim) = copysign(1.0, source_r_rel(i_dim));
			delta_i = abs(test_i - source_i);
			int adjacentness = sum(delta_i);
			if(adjacentness == 1) {
				int i_dim_hit = argmax(delta_i);
				i_dims_hit = Range(i_dim_hit, i_dim_hit);
			} else if(adjacentness == 2) {
				v_a = abs(m_vs(i_arrow, Range::all()));
				int i_dim_bias = argmax(v_a);
				pos_i = test_i;
				pos_i(1 - i_dim_bias) += sides(1 - i_dim_bias);
				if(not box.is_wall(pos_i)) {
					int i_dim_hit = 1 - i_dim_bias;
					i_dims_hit = Range(i_dim_hit, i_dim_hit);
				} else {
					int i_dim_nonbias = 1 - i_dim_bias;
					pos_i = test_i;
					pos_i(i_dim_nonbias) += sides(1 - i_dim_nonbias);
					if (not box.is_wall(pos_i)) {
						int i_dim_hit = 1 - i_dim_nonbias;
						i_dims_hit = Range(i_dim_hit, i_dim_hit);
					} else {
//						i_dims_hit = Range::all();
						i_dims_hit = Range(0, 1);
					}
				}
			} else {
				cout << "Error: Update rs went wrong (invalid adjacentness: " << adjacentness << " )." << endl;
			}
			test_r(i_dims_hit) = test_cell_r(i_dims_hit) + sides(i_dims_hit) *
				((box.dx_get() / 2.0) + box.cell_buffer_get());
			wall_specular(i_arrow, i_dims_hit);
		}
		m_rs(i_arrow, Range::all()) = test_r;
	}
}

void Arrows::vs_update(void) {
	tumble();
}

void Arrows::rates_update(const Box& box) {
	(this->*m_prates_update)(box);
}

void Arrows::wall_handle(int i_arrow, Range i_dims_hit) {
	(this->*m_pwall_handle)(i_arrow, i_dims_hit);
}

const Array<float, DIM>& Arrows::rs_get(void) const {
	return m_rs;
}

const Array<float, DIM>& Arrows::vs_get(void) const {
	return m_vs;
}

int Arrows::num_arrows_get(void) const {
	return m_num_arrows;
}

void Arrows::rs_initialise(const Box& box) {
	m_rs.resize(m_num_arrows, DIM);
	float L_half = box.L_get() / 2.0;
	for(int i_arrow = 0; i_arrow < m_num_arrows; i_arrow++) {
		for(int i_dim = 0; i_dim < DIM; i_dim++)
			m_rs(i_arrow, i_dim) = random_float(-L_half, +L_half);
		while(box.is_wall(m_rs(i_arrow, Range::all())))
			for(int i_dim = 0; i_dim < DIM; i_dim++)
				m_rs(i_arrow, i_dim) = random_float(-L_half, +L_half);
	}
}

void Arrows::vs_initialise() {
	m_vs.resize(m_num_arrows, DIM);
	Array<float, 1> v_p(DIM);
	v_p(0) = m_v;
	for(int i_arrow = 0; i_arrow < m_num_arrows; i_arrow++) {
		v_p(1) = random_float(-M_PI, +M_PI);
		m_vs(i_arrow, Range::all()) = polar_to_cart(v_p);
	}
}

void Arrows::rates_initialise(char rate_alg, float sense_mem, float t_mem) {
	m_rates.resize(m_num_arrows);

	if (rate_alg == 'm') {
		float A = 0.5;
		float N = pow((0.8125 * pow(A, 2) - 0.75 * A + 0.5), -0.5);
		int i_t__max = static_cast<int>(t_mem / DELTA_t);
		Array<float, 1> g(i_t__max);
		g = m_rate_base * i1 * DELTA_t;
		m_K.resize(g.shape());
		m_K = N * sense_mem * m_rate_base * exp(-g) * (1 - A * (g + pow2(g) / 2.0));
		m_attract_mem.resize(m_num_arrows, i_t__max);
	}
}

void Arrows::tumble(void) {
	Array<float, 1> v_p(DIM), v_c(DIM);
	for(int i_arrow = 0; i_arrow < m_num_arrows; i_arrow++) {
		float dice_roll = random_float(0.0, 1.0);
		if(dice_roll < (m_rates(i_arrow) * DELTA_t)) {
			v_c = m_vs(i_arrow, Range::all());
			v_p = cart_to_polar(v_c);
			v_p(1) = random_float(-M_PI, +M_PI);
			m_vs(i_arrow, Range::all()) = polar_to_cart(v_p);
		}
	}
}

float Arrows::grad_find(const Box& box, Array<float, DIM>& field, int i_arrow) {
	TinyVector<int, DIM> arrow_i = box.r_to_i(m_rs(i_arrow, Range::all()));

	Array<float, 1> v_abs(abs(m_vs(i_arrow, Range::all())));
	int i_dim_bias = argmax(v_abs);
	int bias_sign = static_cast<int>(copysign(1.0, m_vs(i_arrow, i_dim_bias)));

	TinyVector<int, DIM> pos_i(arrow_i);
	pos_i(i_dim_bias) = arrow_i(i_dim_bias) + bias_sign;
	float grad = field(pos_i(0), pos_i(1));
	pos_i(i_dim_bias) = arrow_i(i_dim_bias) - bias_sign;
	grad -= field(pos_i(0), pos_i(1));
	return grad / box.dx_get();
}

float Arrows::integral_find(int i_arrow) {
	return sum(m_attract_mem(i_arrow, Range::all()) * m_K * DELTA_t);
}

void Arrows::rates_update_const(const Box& box) {
		m_rates = m_rate_base;
}

void Arrows::rates_update_grad(const Box& box) {
	Array<float, DIM> attract = box.attract_get();
	for(int i_arrow = 0; i_arrow < m_num_arrows; i_arrow++) {
		m_rates(i_arrow) =
				m_rate_base * (1.0 -
				m_sense_grad * grad_find(box, attract, i_arrow));
	}
}

void Arrows::rates_update_mem(const Box& box) {
	TinyVector<int, DIM> arrow_i;
	const Array<float, DIM>& attract = box.attract_get();
	m_attract_mem(Range::all(), Range(1, toEnd)) =
			m_attract_mem(Range::all(), Range(fromStart, m_attract_mem.extent(secondDim) - 2));
	for(int i_arrow = 0; i_arrow < m_num_arrows; i_arrow++) {
		arrow_i = box.r_to_i(m_rs(i_arrow, Range::all()));
		m_attract_mem(i_arrow, 0) = attract(arrow_i);
		m_rates(i_arrow) = m_rate_base * (1.0 - integral_find(i_arrow));
	}
}

void Arrows::wall_specular(int i_arrow, Range i_dims_hit) {
	m_vs(i_arrow, i_dims_hit) *= -1.0;
}

void Arrows::wall_bounceback(int i_arrow, Range i_dims_hit) {
	m_vs(i_arrow, Range::all()) *= -1.0;
}

void Arrows::wall_aligning(int i_arrow, Range i_dims_hit) {
	Array<float, 1> direction(DIM), v(DIM);
	direction(i_dims_hit) = 0.0;
	float d_mag = vector_mag(direction);
	if(d_mag < ZERO_THRESH) {
		int sign = copysign(1.0, random_float(-1.0, +1.0));
		direction = m_vs(i_arrow, Range::all()).copy();
		for (int i_dim = 0; i_dim < DIM; i_dim++)
			m_vs(i_arrow, i_dim) = sign * direction(1 - i_dim);
	} else {
		v = m_vs(i_arrow, Range::all());
		float v_mag = vector_mag(v);
		m_vs(i_arrow, Range::all()) = direction * (v_mag / d_mag);
	}
}

void Arrows::wall_stalling(int i_arrow, Range i_dims_hit) {
	m_vs(i_arrow, i_dims_hit) = 0.0;
}
